tests/generative/test_query_model.py

```diff
@@ -0,0 +1,357 @@
+    # version discards fewer invalid examples and is noticeably faster, for reasons that I can't understand.
+    #
+    #     wrapped.filter(predicate)
+    example = draw(wrapped)
+    hyp.assume(predicate(example))
+    return example
+
+
+# Here follows a set of recursive strategies for constructing valid query model graphs.
+#
+# There is a set of strategies corresponding to the abstract types in the query model (for example `frame` or
+# `one_row_per_patient_series`), each of which returns one of the direct subtypes of that type (both abstract and
+# concrete). Another set of strategies corresponds to the concrete types in the query model (for example `filter` or
+# `select_column`), each of which constructs the corresponding node and recursively invokes other strategies to
+# provide the constructor arguments (using the strategy corresponding to the types of the arguments).
+#
+# So each concrete node type appears in two places: in its own strategy and in the strategy of its immediate
+# supertype. And each abstract type appears in its own strategy, in the strategy of its immediate supertype and in
+# the strategies of any concrete nodes which have a constructor argument of that abstract type.
+#
+# When we invoke the strategy of an abstract type, it tugs at a thread that runs down through the type hierarchy
+# (picking a random subclass at each level) until it reaches a concrete type which can be constructed. (And it then
+# does the same for all of the necessary constructor arguments.)
+#
+# The strategies are ordered in this file in the same order that the corresponding classes are defined in
+# `query_model`, to make it easier to see the correspondence and to spot errors.
+#
+# There are, you will be astonished to learn, a few complications.
+#
+# * In order to help Hypothesis shrink the examples, the subclasses of each abstract type are ordered roughly by
+#   "simplicity", using the number of child nodes as a heuristic. It's hard to follow even that, though, because of
+#   the way the concrete types are grouped. And, partly as a consequence, this strategy doesn't shrink very well.
+# * We need to use `st.deferred` in some places to allow (transitively) mutually recursive strategies. We're using it
+#   in _all_ the abstract type strategies in order to allow us to keep the ordering the same as that of the definitions
+#   in `query_model`.
+# * The query model type hierarchy doesn't give us all the information we need, so the approach outlined above will
+#   generate invalid query model graphs (due either to the domains or the types being incorrect). We handle this by
+#   constructing the query model objects using a special-purpose strategy `qm_builds`, which can discard invalid
+#   instances. There is an unexplored optimization avenue here: make the strategies more specific to include type
+#   and/or domain, so that we don't discard so many examples.
```

Comment:
I can follow how all this works, and I'm reasonably confident that I could add support for a new kind of QM object, but I'm not sure I fully understand why we need two kinds of strategy.  Is that something you're able to explain in this comment?

Reply:
By two kinds, do you mean the ones for the abstract vs concrete types?

Reply:
@inglesp I'm going to merge this to get it off my plate, but please feel free to clarify what you meant here and I'll try to improve the comments later.